### PR TITLE
Feature/new precise upload

### PR DESCRIPTION
**Pull request by chrisveilleux**
## Description
Change logic that uploads wake word samples to use the new endpoint residing in Selene

## How to test
Opt in to open dataset, speak the wake word, see the sample uploaded to the Selene Device API server.

## Contributor license agreement signed?
CLA [Y]

### CODE REVIEW

**Comment by devops-mycroft** [comment]
Voight Kampff Integration Test Succeeded  ([Results](https://reports.mycroft.ai/core/PR-2753))

File: mycroft/client/speech/mic.py [hunk 1]
```diff
@@ -513,32 +520,24 @@ def _compile_metadata(self):
...
-            "Mycroft-core v20.2 and below has been deprecated. To contribute "
-            "new wakeword samples please upgrade to v20.8 or above."
-        )
-        # def upload(audio, metadata):
-        #     requests.post(self.upload_url,
-        #                   files={'audio': BytesIO(audio.get_wav_data()),
-        #                          'metadata': StringIO(json.dumps(metadata))})
-        # Thread(target=upload, daemon=True, args=(audio, metadata)).start()
+        """Upload the wake word to the backend servers."""
+        self.device_api.upload_wake_word(audio, metadata)
```
Comment (forslund) [inline mycroft/client/speech/mic.py:540]: This was previously done by launching a background thread to not block. Should that be used now as well?

File: mycroft/api/__init__.py [hunk 2]
```diff
@@ -424,6 +426,29 @@ def upload_skills_data(self, data):
             "json": to_send
             })
 
+    def upload_wake_word(self, audio, params):
+        """Helper method for sending files using the device API"""
```
Comment (forslund) [inline mycroft/api/__init__.py:430]: Nitpick: The `audio` and `params` parameters should be documented in the docstring.

**Review by forslund — CHANGES_REQUESTED** [review]
I get this exception

```
2020-11-09 22:24:53.372 | ERROR    | 19415 | mycroft.client.speech.listener:run:94 | IOError Exception in AudioProducer
Traceback (most recent call last):
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 603, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 362, in _error_catcher
    yield
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 668, in read_chunked
    self._update_chunk_length()
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 607, in _update_chunk_length
    raise httplib.IncompleteRead(line)
http.client.IncompleteRead: IncompleteRead(0 bytes read)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/models.py", line 750, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 492, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 696, in read_chunked
    self._original_response.close()
  File "/usr/local/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/urllib3/response.py", line 380, in _error_catcher
    raise ProtocolError('Connection broken: %r' % e, e)
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ake/projects/python/mycroft-core/mycroft/client/speech/listener.py", line 84, in run
    self.stream_handler)
  File "/home/ake/projects/python/mycroft-core/mycroft/client/speech/mic.py", line 716, in listen
    self._handle_wakeword_found(ww_data.audio, source)
  File "/home/ake/projects/python/mycroft-core/mycroft/client/speech/mic.py", line 583, in _handle_wakeword_found
    self._upload_wakeword(audio, metadata)
  File "/home/ake/projects/python/mycroft-core/mycroft/client/speech/mic.py", line 540, in _upload_wakeword
    self.device_api.upload_wake_word(audio, metadata)
  File "/home/ake/projects/python/mycroft-core/mycroft/api/__init__.py", line 446, in upload_wake_word
    'metadata': StringIO(json.dumps(request_data))
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/api.py", line 116, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/sessions.py", line 524, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/sessions.py", line 677, in send
    r.content
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/models.py", line 828, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
  File "/home/ake/projects/python/mycroft-core/.venv/lib/python3.7/site-packages/requests/models.py", line 753, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(0 bytes read)', IncompleteRead(0 bytes read))
```

So something's not working? Or do I need to be on the test backend or something?

In addition I think we should try to catch the possible exceptions here so a bad connection or server-response doesn't block the flow here (even if it's moved back into a thread I think it's a good idea to handle the errors).

I also made a couple of minor notes in the diff, but in general (and apart from the exception it looks very nice and clean)

**Comment by forslund** [comment]
What is the status of this PR? is it still something that should be implemented?

**Comment by chrisveilleux** [comment]
> What is the status of this PR? is it still something that should be implemented?

I believe this is meant to replace an endpoint in the Precise API that is deprecated.  So yes, it still needs to be implemented.

**Comment by forslund** [comment]
Closing PR since we're archiving the repo

**Pull request closed without merging**